numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1505,62 +1495,47 @@ pack_inner(const char *inptr,
+#if NPY_SIMD
     if (in_stride == 1 && element_size == 1 && n_out > 2) {
-        __m128i zero = _mm_setzero_si128();
+        npyv_u8 v_zero = npyv_zero_u8();
         /* don't handle non-full 8-byte remainder */
         npy_intp vn_out = n_out - (remain ? 1 : 0);
-        vn_out -= (vn_out & 1);
-        for (index = 0; index < vn_out; index += 2) {
-            unsigned int r;
-            npy_uint64 a = *(npy_uint64*)inptr;
-            npy_uint64 b = *(npy_uint64*)(inptr + 8);
-            if (order == 'b') {
-                a = npy_bswap8(a);
-                b = npy_bswap8(b);
+        const int vstep = npyv_nlanes_u64;
+        const int vstepx4 = vstep * 4;
+        vn_out -= (vn_out & (vstep - 1));
+        for (; index <= vn_out - vstepx4; index += vstepx4, inptr += npyv_nlanes_u8 * 4) {
+            npyv_u8 v0 = npyv_load_u8((const npy_uint8*)inptr);
+            npyv_u8 v1 = npyv_load_u8((const npy_uint8*)inptr + npyv_nlanes_u8 * 1);
+            npyv_u8 v2 = npyv_load_u8((const npy_uint8*)inptr + npyv_nlanes_u8 * 2);
+            npyv_u8 v3 = npyv_load_u8((const npy_uint8*)inptr + npyv_nlanes_u8 * 3);
+            if (order == PACK_ORDER_BIG) {
+                v0 = npyv_rev64_u8(v0);
+                v1 = npyv_rev64_u8(v1);
+                v2 = npyv_rev64_u8(v2);
+                v3 = npyv_rev64_u8(v3);
             }
-            
-            /* note x86 can load unaligned */
-            __m128i v = _mm_set_epi64(_m_from_int64(b), _m_from_int64(a));
-            /* false -> 0x00 and true -> 0xFF (there is no cmpneq) */
-            v = _mm_cmpeq_epi8(v, zero);
-            v = _mm_cmpeq_epi8(v, zero);
-            /* extract msb of 16 bytes and pack it into 16 bit */
-            r = _mm_movemask_epi8(v);
-            /* store result */
-            memcpy(outptr, &r, 1);
-            outptr += out_stride;
-            memcpy(outptr, (char*)&r + 1, 1);
-            outptr += out_stride;
-            inptr += 16;
-        }
-    }
-#elif defined NPY_HAVE_NEON
-    if (in_stride == 1 && element_size == 1 && n_out > 2) {
-        /* don't handle non-full 8-byte remainder */
-        npy_intp vn_out = n_out - (remain ? 1 : 0);
-        vn_out -= (vn_out & 1);
-        for (index = 0; index < vn_out; index += 2) {
-            unsigned int r;
-            npy_uint64 a = *((uint64_unaligned*)inptr);
-            npy_uint64 b = *((uint64_unaligned*)(inptr + 8));
-            if (order == 'b') {
-                a = npy_bswap8(a);
-                b = npy_bswap8(b);
+            npy_uint64 bb[4];
+            bb[0] = npyv_tobits_b8(npyv_cmpneq_u8(v0, v_zero));
+            bb[1] = npyv_tobits_b8(npyv_cmpneq_u8(v1, v_zero));
+            bb[2] = npyv_tobits_b8(npyv_cmpneq_u8(v2, v_zero));
+            bb[3] = npyv_tobits_b8(npyv_cmpneq_u8(v3, v_zero));
+            for(int i = 0; i < 4; i++) {
+                for (int j = 0; j < vstep; j++) {
+                    memcpy(outptr, (char*)&bb[i] + j, 1);
```

Comment:
Would be better if you get rid of `memcpy` if out array is contiguous `out_stride == 1`
by packing the bits before storing them?

```C
// you will need to check if outptr alignment by 8-byte on arm7 to avoid bus error
npy_uint64 *ptr64 = (npy_uint64*)outptr;
#if NPY_SIMD_WIDTH == 16
    npy_uint64 bb = bb0 | (bb1 << 16) | (bb2 << 32) | (bb3 << 48);
    ptr64[0] = bb;
#elif NPY_SIMD_WIDTH == 32
    ptr64[0] = bb0 | (bb1 << 32);
    ptr64[1] = bb2 | (bb3 << 32);
#else
    ptr64[0] = bb0; ptr64[1] = bb1;
    ptr64[2] = bb2; ptr64[3] = bb3;
#endif
```

Reply:
Other architect may have bus error too, so 8-bytes alignment maybe necessary.

Reply:
I don't think so, just armv7 and below. could you add a C #def for it and using it as a guard? similar to the OpenCV one

https://github.com/opencv/opencv/blob/46e275dfe4616b07cd2492c17698d1a0be0aa6df/modules/core/include/opencv2/core/cvdef.h#L345-L351

Reply:
The defination has added.

Reply:
Would it be better to follow OpenCV approach? we may face this issue in other areas.
so I suggest adding a new #def `NPY_STRONG_ALIGNMENT` to `numpy/npy_common.h` similar to the OpenCV one.

Reply:
you are right, The alignment attribute is widely used.